ext/standard/string.c

```diff
@@ -1866,6 +1866,42 @@ PHP_FUNCTION(str_contains)
 }
 /* }}} */
 
+/* {{{ proto bool str_starts_with(string haystack, string needle)
+   Checks if haystack starts with needle */
+PHP_FUNCTION(str_starts_with) {
+    zend_string *haystack, *needle;
+
+    ZEND_PARSE_PARAMETERS_START(2, 2)
+        Z_PARAM_STR(haystack)
+        Z_PARAM_STR(needle)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (ZSTR_LEN(needle) > ZSTR_LEN(haystack)) {
+        RETURN_FALSE;
+    }
+
+    RETURN_BOOL(memcmp(ZSTR_VAL(haystack), ZSTR_VAL(needle), ZSTR_LEN(needle)) == 0);
+}
+
+/* {{{ proto bool str_ends_with(string haystack, string needle)
+   Checks if haystack ends with needle */
+PHP_FUNCTION(str_ends_with) {
+    zend_string *haystack, *needle;
+    int k;
+
+    ZEND_PARSE_PARAMETERS_START(2, 2)
+        Z_PARAM_STR(haystack)
+        Z_PARAM_STR(needle)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (ZSTR_LEN(needle) > ZSTR_LEN(haystack)) {
+        RETURN_FALSE;
+    }
+
+    k = ZSTR_LEN(haystack) - ZSTR_LEN(needle);
+    RETURN_BOOL(memcmp(&(ZSTR_VAL(haystack))[k], ZSTR_VAL(needle), ZSTR_LEN(needle)) == 0);
```

Comment:
`&(ZSTR_VAL(haystack))[k]` seems "non idiomatic" in this file, I would rather see
`ZSTR_VAL(haystack) + ZSTR_LEN(haystack) - ZSTR_LEN(needle)` and get rid of `k` entirely
(but let's see what nikic thinks)

Reply:
I'm alright with changing it, I don't do a lot of C so I had to play around some to figure that one out. More than happy to use a more concise or idiomatic way.